crates/ra_hir/src/ids.rs

```diff
@@ -130,6 +133,105 @@ impl MacroCallLoc {
     }
 }
 
+#[derive(Debug, PartialEq, Eq, Hash)]
+pub struct ItemLoc<N: AstNode> {
+    pub(crate) module: Module,
+    raw: SourceItemId,
+    _ty: PhantomData<N>,
+}
+
+impl<N: AstNode> ItemLoc<N> {
+    pub(crate) fn from_ast(
+        db: &impl HirDatabase,
+        module: Module,
+        file_id: HirFileId,
+        ast: &N,
+    ) -> ItemLoc<N> {
+        let items = db.file_items(file_id);
+        let raw = SourceItemId {
+            file_id,
+            item_id: Some(items.id_of(file_id, ast.syntax())),
+        };
+        ItemLoc {
+            module,
+            raw,
+            _ty: PhantomData,
+        }
+    }
+
+    pub(crate) fn source(&self, db: &impl HirDatabase) -> (HirFileId, TreeArc<N>) {
+        let syntax = db.file_item(self.raw);
+        let ast = N::cast(&syntax)
+            .unwrap_or_else(|| panic!("invalid ItemLoc: {:?}", self.raw))
+            .to_owned();
+        (self.raw.file_id, ast)
+    }
+}
+
+impl<N: AstNode> Clone for ItemLoc<N> {
+    fn clone(&self) -> ItemLoc<N> {
+        ItemLoc {
+            module: self.module,
+            raw: self.raw,
+            _ty: PhantomData,
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub struct FunctionId(RawId);
+impl_arena_id!(FunctionId);
+
+pub(crate) type FunctionLoc = ItemLoc<ast::FnDef>;
+
+impl FunctionId {
+    pub(crate) fn loc(self, db: &impl AsRef<HirInterner>) -> FunctionLoc {
+        db.as_ref().fns.id2loc(self)
+    }
+}
+
+impl FunctionLoc {
+    pub(crate) fn id(&self, db: &impl AsRef<HirInterner>) -> FunctionId {
+        db.as_ref().fns.loc2id(&self)
+    }
+}
+
```

Comment:
I'm not a big fan of macros but maybe this would be a good place to use them to reduce the boilerplate.

Reply:
should be better now.